### PR TITLE
[FIXED JENKINS-43856] Restore binary compatibility with pre-2.6 releases

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-43856](https://issues.jenkins-ci.org/browse/JENKINS-43856)

cc @reviewbybees and @i386

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java [hunk 1]
```diff
@@ -165,30 +165,31 @@ public HttpResponse doProceed(StaplerRequest request) throws IOException, Servle
...
             listener.getLogger().println("Approved by " + hudson.console.ModelHyperlinkNote.encodeTo(user));
         }
-        node.addAction(new InputSubmittedAction(approverId, params));
-
-        Object v;
-        if (params != null && params.size() == 1) {
-            v = params.values().iterator().next();
-        } else {
-            v = params;
+        if (v != null && v instanceof Map) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java]: This is not a good idea.

Revert. Have the `Map` overload be the real implementation. Restore the original method overload with `Object`, deprecate it, and have it delegate to the new overload.

Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java]: Duh, that's smarter. Thanks. =)

Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java]: Updated.

**Comment by i386** [comment]
If you can manually confirm it works with Blue Ocean 1.0.1 then 🐝

Thanks for jumping in here so quickly

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java [hunk 2]
```diff
@@ -193,6 +193,19 @@ public HttpResponse proceed(@CheckForNull Map<String,Object> params) {
         return HttpResponses.ok();
     }
 
+    @Deprecated
+    @SuppressWarnings("unchecked")
+    public HttpResponse proceed(Object v) {
+        Map<String,Object> params = new HashMap<>();
+        if (v instanceof Map) {
+            params.putAll((Map)v);
+        } else {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java]:
```java
} else if (v != null) {
```

Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java]: Good catch.

**Comment by abayer** [comment]
@i386 Tested first with `pipeline-input-step` 2.6 and `blueocean` 1.0.1, and it failed. Tested again with SNAPSHOT of this PR and `blueocean` 1.0.1 and it worked. So presumably good. =)

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java [hunk 3]
```diff
@@ -188,11 +188,26 @@ public HttpResponse proceed(@CheckForNull Map<String,Object> params) {
...
+            params.putAll((Map) v);
+        } else if (v != null) {
+            params.put("parameter", v);
+        }
+
+        if (params.isEmpty()) {
+            return proceed(null);
+        } else {
+            return proceed(params);
+        }
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java]: More simply:

```java
if (v instanceof Map) {
    return proceed(new HashMap<String,Object>((Map) v);
} else if (v == null) {
    return proceed(null);
} else {
    return proceed(Collections.singletonMap("parameter", v));
}
```

Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputStepExecution.java]: Nice, doing so.

**Comment by i386** [comment]
Ship it :)

**Comment by michaelneale** [comment]
:bee: nice catch. GO!

**Comment by i386** [comment]
Thanks @jglick and @abayer :)

**Pull request merged**